cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,297 @@
+
+  public ClusteredSortAgentScheduler(JedisSource jedisSource, NodeStatusProvider nodeStatusProvider, AgentIntervalProvider intervalProvider, Integer parallelism) {
+    this.jedisSource = jedisSource;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agents = new ConcurrentHashMap<>();
+    this.intervalProvider = intervalProvider;
+    this.log = LoggerFactory.getLogger(getClass());
+
+    if (parallelism == 0 || parallelism < -1) {
+      throw new IllegalArgumentException("Argument 'parallelism' must be positive, or -1 (for unlimited parallelism).");
+    } else if (parallelism > 0) {
+      this.runningAgents = Optional.of(new Semaphore(parallelism));
+    } else {
+      this.runningAgents = Optional.empty();
+    }
+
+    scriptShas = new ConcurrentHashMap<>();
+    storeScripts();
+
+    this.agentWorkPool = Executors.newCachedThreadPool(new NamedThreadFactory(AgentWorker.class.getSimpleName()));
+    Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(ClusteredSortAgentScheduler.class.getSimpleName()))
+      .scheduleAtFixedRate(this, 0, 1, TimeUnit.SECONDS);
+  }
+
+  private void storeScripts() {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      // When we switch an agent from one set to another, we first make sure it exists in the set we are removing it
+      // from, and then we perform the swap. If this check fails, the thread performing the swap does not get ownership
+      // of the agent.
+      // Swap happens from KEYS[1] -> KEYS[2] with the agent type being ARGV[1], and the score being ARGV[2].
+      scriptShas.put(SWAP_SET_SCRIPT, jedis.scriptLoad(
+          "if redis.call('zrank', KEYS[1], ARGV[1]) ~= nil then\n" +
+          "  redis.call('zrem', KEYS[1], ARGV[1])\n" +
+          "  return redis.call('zadd', KEYS[2], ARGV[2], ARGV[1])\n" +
+          "else return nil end\n"
+      ));
+
+      scriptShas.put(CONDITIONAL_SWAP_SET_SCRIPT, jedis.scriptLoad(
+          "if redis.call('zrank', KEYS[1], ARGV[1]) == ARGV[3] then\n" +
+          "  redis.call('zrem', KEYS[1], ARGV[1])\n" +
+          "  return redis.call('zadd', KEYS[2], ARGV[2], ARGV[1])\n" +
+          "else return nil end\n"
+      ));
+
+      // If the agent isn't present in either the WAITING or WORKING sets, it's safe to add. If it's present in either,
+      // it's being worked on or was recently run, so leave it be.
+      // KEYS[1] and KEYS[2] are checked for inclusion. If the agent is in neither ARGV[1] is added to KEYS[1] with score
+      // ARGV[2].
+      scriptShas.put(ADD_AGENT_SCRIPT, jedis.scriptLoad(
+        "if redis.call('zrank', KEYS[1], ARGV[1]) ~= nil then\n" +
+        "  if redis.call('zrank', KEYS[2], ARGV[1]) ~= nil then\n" +
+        "    return redis.call('zadd', KEYS[1], ARGV[2], ARGV[1])\n" +
+        "  else return nil end\n" +
+        "else return nil end\n"));
+    }
+  }
+
+  private String getScriptSha(String scriptName) {
+    try (Jedis jedis = jedisSource.getJedis()) {
```

Comment:
move the try-with-resources bit to just contain the call to jedis (otherwise it has a connection from the pool, then gets another connection from the pool inside storeScripts) Alternatively have storeScripts accept the Jedis object as an arg?